src/selectors/linked-nodes.js

```diff
@@ -17,36 +15,83 @@ export const getCentralNode = createSelector(
 );
 
 /**
- * Recursively search through the edges data for ancestor and descendant nodes
+ * Gets a map of nodeIDs to the visible source edges and target edges
  * @param {Array} edges
+ */
+export const getVisibleEdgesByNode = createSelector(
+  [getVisibleEdges],
+  edges => {
+    const sourceEdges = {};
+    const targetEdges = {};
+
+    for (const edge of edges) {
+      if (!sourceEdges[edge.target]) {
+        sourceEdges[edge.target] = [];
+      }
+
+      sourceEdges[edge.target].push(edge);
+
+      if (!targetEdges[edge.source]) {
+        targetEdges[edge.source] = [];
+      }
+
+      targetEdges[edge.source].push(edge);
+    }
+
+    return { sourceEdges, targetEdges };
+  }
+);
+
+/**
+ * Finds all visible successors for the given nodeID in given direction
+ * @param {string} nodeID
+ * @param {Object} edgesByNode
+ * @param {string} direction
+ * @param {?object} visited
+ * @param {?boolean} isStart
+ * @param {?Array} empty
+ */
+const findLinkedNodes = (
+  nodeId,
+  edgesByNode,
+  direction,
+  visited = {},
+  isStart = true,
+  empty = []
+) => {
+  if (isStart || !visited[nodeId]) {
+    visited[nodeId] = true;
+
+    (edgesByNode[nodeId] || empty).forEach(edge =>
+      findLinkedNodes(
+        edge[direction],
+        edgesByNode,
+        direction,
+        visited,
+        false,
+        empty
+      )
+    );
+  }
+
+  return visited;
+};
```

Comment:
I'd prefer to avoid this many arguments, because I find more than three a little confusing to read. 

Personally I'd recommend currying instead and placing most of the variables (e.g. `edgesByNode`, `direction`, `visited`, `empty`) in the parent scope, because they are consistent for each run. e.g. something like

```es6
const findLinkedNodes = (edgesByNode, direction, visited) => {
  let empty; //...
  return (nodeId) => (function(nodeID){ /*...*/ })(nodeID);
};

const findLinkedSourceNodes = findLinkedNodes(sourceEdges, 'source', linkedNodes);
findLinkedSourceNodes(nodeID);
```

Else, I'd use an object so that they're named.

But this is mostly a personal preference thing, so I'm not gonna block this PR because of it.

Reply:
I see what you're saying, I've now added comments, removed the empty array and generalised `isStart` to `depth` - hopefully all of that makes it much clearer?